internal/wasm/memory.go

```diff
@@ -279,3 +297,109 @@ func (m *MemoryInstance) writeUint64Le(offset uint32, v uint64) bool {
 	binary.LittleEndian.PutUint64(m.Buffer[offset:], v)
 	return true
 }
+
+// Wait32 suspends the caller until the offset is notified by a different agent.
+func (m *MemoryInstance) Wait32(offset uint32, exp uint32, timeout int64) uint64 {
+	w := m.getWaiters(offset)
+	w.mux.Lock()
+
+	addr := unsafe.Add(unsafe.Pointer(&m.Buffer[0]), offset)
+	// Must be atomic even though we have the lock since it can be modified
+	// concurrently by a non-wait instruction.
+	cur := atomic.LoadUint32((*uint32)(addr))
```

Comment:
Is your expectation that unlocking `MemoryInstance.Mux` makes those changes visible here?

Reply:
I see what you mean let me fix

Reply:
I'm actually not sure; as since we're assuming for fence that a mutex implies a barrier/fence, maybe this works (under the same assumption).

If that's the rational, I'd document it, though (just as you did on fence).

Reply:
Thanks, I have updated the comment. I do think mux unlock will make changes visible to future atomic reads. I did think that perhaps there is still an issue that we won't read the value from an atomic write whose mux has not been unlocked yet - does that seem OK? Otherwise we would need to acquire the lock in interpreter mode, and unlock after reading this value, before potentially blocking. It makes the code more hairy but I wonder if it's required, what do you think?

Reply:
I need a functioning brain to answer that.
Let me try again tomorrow.

Reply:
Thanks @ncruces - I went ahead and pushed a commit too that while a bit annoying, isn't so bad, and it makes sure cur value is read in interpreter in the same way as other memory operations. We can revert if it is seems unnecessary

Reply:
BTW this may be useful reading

https://github.com/WebAssembly/threads/issues/211

Other alternatives could be simpler code if the current value read didn't need to be under the waiters lock

Reply:
LGTM